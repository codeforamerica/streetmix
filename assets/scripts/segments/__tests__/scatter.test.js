/* eslint-env jest */
import { getRandomObjects } from '../scatter'
import PEOPLE from '../people.json'

// The unit test for `getRandomObjects()` is designed to ensure that it
// returns the same values as previous behaviour, because we're heavily
// refactoring how this works, and want to avoid regressions. We will need to
// continue making sure that we return similar output against future
// refactoring -- for instance, when transitioning from imperial to metric
// units.

// However! The output of this function is not intended to be guaranteed
// consistent across all releases, because it can change for any reason -- for
// example, just adding more people to people.json will result in different
// output. As a future improvement, we can either make this test more general,
// or roll it up into an integration test.
describe('scatter objects in segments', () => {
  it('picks people at normal density', () => {
    const [people, startLeft] = getRandomObjects(
      PEOPLE,
      28.66667,
      936877893,
      0,
      5,
      4,
      -0.5
    )

    expect(people).toMatchSnapshot()
    expect(startLeft).toEqual(3.7987234759633886)
  })

  it('picks people at sparse density', () => {
    const [people, startLeft] = getRandomObjects(
      PEOPLE,
      28.66667,
      936877893,
      3.5,
      11.8333,
      4,
      -0.5
    )

    expect(people).toMatchSnapshot()
    expect(startLeft).toEqual(4.846658384686)
  })

  it('picks people at dense density', () => {
    const [people, startLeft] = getRandomObjects(
      PEOPLE,
      28.66667,
      936877893,
      0,
      1.5,
      4,
      -0.5
    )

    expect(people).toMatchSnapshot()
    expect(startLeft).toEqual(3.0462522570605604)
  })

  it('picks from a pool of one single string id', () => {
    const scatter = {
      sprites: ['plants--bush'],
      minSpacing: -3,
      maxSpacing: 3
    }
    const pool = [
      {
        id: 'plants--bush',
        width: 5
      }
    ]

    const [objects, startLeft] = getRandomObjects(
      pool,
      20,
      0,
      scatter.minSpacing,
      scatter.maxSpacing,
      5,
      0
    )

    expect(objects).toMatchSnapshot()
    expect(startLeft).toEqual(4.430707699360696)
  })
})
