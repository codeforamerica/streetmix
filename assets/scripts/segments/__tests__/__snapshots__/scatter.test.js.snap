// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scatter objects in segments picks from a pool of one single string id 1`] = `
Array [
  Object {
    "id": "plants--bush",
    "left": 0,
    "width": 5,
  },
  Object {
    "id": "plants--bush",
    "left": 2.5174780997211976,
    "width": 5,
  },
  Object {
    "id": "plants--bush",
    "left": 7.687586461069147,
    "width": 5,
  },
  Object {
    "id": "plants--bush",
    "left": 11.138584601278607,
    "width": 5,
  },
]
`;

exports[`scatter objects in segments picks people at dense density 1`] = `
Array [
  Object {
    "id": "people-23",
    "left": 0,
    "name": "Katie",
    "width": 3,
  },
  Object {
    "id": "people-28",
    "left": 3.7407393094867594,
    "width": 3,
  },
  Object {
    "id": "people-21",
    "left": 6.615210042353208,
    "width": 1,
  },
  Object {
    "id": "people-22",
    "left": 7.860021493652412,
    "width": 1,
  },
  Object {
    "id": "people-25",
    "left": 9.368637363837973,
    "width": 2,
  },
  Object {
    "id": "people-12",
    "left": 11.270766914421966,
    "width": 2,
  },
  Object {
    "id": "people-24",
    "left": 13.983675515088438,
    "width": 4,
  },
  Object {
    "id": "people-15",
    "left": 18.618860351446976,
    "width": 2,
  },
  Object {
    "id": "people-29",
    "left": 21.074165485878872,
    "width": 4,
  },
]
`;

exports[`scatter objects in segments picks people at normal density 1`] = `
Array [
  Object {
    "id": "people-23",
    "left": 0,
    "name": "Katie",
    "width": 3,
  },
  Object {
    "id": "people-28",
    "left": 6.6357976982892,
    "width": 3,
  },
  Object {
    "id": "people-21",
    "left": 10.384033474510698,
    "width": 1,
  },
  Object {
    "id": "people-22",
    "left": 13.36673831217471,
    "width": 1,
  },
  Object {
    "id": "people-25",
    "left": 17.228791212793247,
    "width": 2,
  },
  Object {
    "id": "people-12",
    "left": 20.069223048073223,
    "width": 2,
  },
]
`;

exports[`scatter objects in segments picks people at sparse density 1`] = `
Array [
  Object {
    "id": "people-23",
    "left": 0,
    "name": "Katie",
    "width": 3,
  },
  Object {
    "id": "people-28",
    "left": 12.892968591830677,
    "width": 3,
  },
  Object {
    "id": "people-21",
    "left": 20.973353230628,
    "width": 1,
  },
]
`;
